### PR TITLE
Eliminate fixed rate loop, block on data receipt

### PR DESCRIPTION
**Pull request by mikepurvis**
It's a bit of a jumble, as I also cleaned up diagnostics and some of initialization/reinitialization works, but the core of this change is that the read loop is now:
- use select() to block on data becoming available (100ms timeout)
- grab the laser's timestamp at this point.
- sleep for 40ms to wait for the buffer to fill up.
- read the contents of the buffer.
- process and publish laserscan.

If the timeout expires in the first step, or there's an fd error, it waits one second and reopens the connection. Recovery and diagnostics look good from testing with a VM on my laptop, with USB connect/disconnect as well as powering the laser down and up.

@servos @abencz to review.

### CODE REVIEW

File: external/SerialDevice/src/SerialDevice.cc [hunk 1]
```diff
@@ -390,6 +390,36 @@ int SerialDevice::WritePort(char *chars, int *written_bytes, int length) {
...
+  FD_ZERO(&rfds);
+  FD_SET(fd, &rfds);
+  tv.tv_sec = 0;
+  tv.tv_usec = 1000 * millis;
+
+  // Wait until we get a return value other than it being interrupted.
+  while(1) {
+    retval = select(fd+1, &rfds, NULL, NULL, &tv);
+    if (retval >= 1) return true;
+    if (retval = 0) return false;
```
Comment (efernandez) [inline external/SerialDevice/src/SerialDevice.cc]: `==` ?!

File: external/SerialDevice/src/SerialDevice.cc [hunk 2]
```diff
@@ -390,6 +390,36 @@ int SerialDevice::WritePort(char *chars, int *written_bytes, int length) {
...
+    if (retval = -1)
```
Comment (efernandez) [inline external/SerialDevice/src/SerialDevice.cc]: == ?!

File: src/s3000_laser.cc [hunk 3]
```diff
@@ -62,30 +59,28 @@ class s3000node {
...
-  diagnostic_updater::FrequencyStatus freq_diag_;
 
+  typedef diagnostic_updater::DiagnosedPublisher<sensor_msgs::LaserScan> LaserScanDiagnosedPublisher;
+  boost::scoped_ptr<LaserScanDiagnosedPublisher> data_pub_;
 
-  s3000node(ros::NodeHandle h) : self_test_(), diagnostic_(),
+  s3000node(ros::NodeHandle h) :
     node_handle_(h), private_node_handle_("~"),
-    desired_freq_(20),
+    desired_freq_(16.7),
```
Comment (efernandez) [inline src/s3000_laser.cc]: In this other part of the code `16.6` is used:
https://github.com/clearpathrobotics/s3000_laser/pull/5/files#diff-3a47006afa665d86bf66c296076b69ebR120

**Comment by efernandez** [comment]
I've just looked a little into the changes and simply commented on some lines of code that might be wrong.

BTW, although it's not related with this PR, I think https://github.com/clearpathrobotics/s3000_laser/pull/5/files#diff-3a47006afa665d86bf66c296076b69ebR123 should be increased or at least analysed, since some values are slightly greater that `49m` (and still below that `52.<something>m`, which seems to be the actual max range). Anyway, this is probably not really important.

File: external/SerialDevice/src/SerialDevice.cc (hunk 2 above)
Comment (mikepurvis) [inline external/SerialDevice/src/SerialDevice.cc]: Embarrassing. Will fix.

File: src/s3000_laser.cc (hunk 3 above)
Comment (mikepurvis) [inline src/s3000_laser.cc]: Didn't realize that; will harmonize that function.

**Comment by servos** [comment]
Other then the few comments @efernandez made, looks good to me.

**Comment by mikepurvis** [comment]
@efernandez Should be better now.

**Comment by mikepurvis** [comment]
Did another quick sanity check on the bot, looked good.

**Comment by efernandez** [comment]
A little late, but :+1:

**Pull request merged**